lib/kafka/client.rb

```diff
@@ -124,7 +126,14 @@ def initialize(seed_brokers:, client_id: "ruby-kafka", logger: nil, connect_time
       )
 
       @cluster = initialize_cluster
-      @partitioner = partitioner || Partitioner.new
+      @partitioner =
+        if partitioner
+          partitioner
+        elsif partitioner_klass
+          Object.const_get(partitioner_klass).new
+        else
+          Partitioner.new
+        end
```

Comment:
I don't think we should introduce a new concept _next to_ `Partitioner`. Instead, `Partitioner` should be configured with the hash/digest algorithm, with the default being `crc32` but `murmur2` being an option. The "random" part would just always be the behavior when no key is present.

So: can we introduce a new parameter to `Partitioner#initialize` called e.g. `hash_function`, which allows either `crc32` (default) or `murmur2`. Clients that wish to use `murmur2` would need to instantiate `Partitioner` with that value and pass the instance to `producer`.

In Racecar and other high-level frameworks, we can add configuration keys for controlling this declaratively.

Reply:
That sounds good. I was taking inspiration from `librdkafka` but your idea will also work well with the opt-in behaviour below.

Just to be clear, you would like the `Partitioner` instance to be passed in, ala [`sasl_oauth_token_provider`](https://github.com/zendesk/ruby-kafka/blob/0d6c4510cb0bd881311d45d3c7cef148f0737805/README.md#oauthbearer), and not the `hash_function` as a symbol ala [`compression_codec`](https://github.com/zendesk/ruby-kafka/blob/0d6c4510cb0bd881311d45d3c7cef148f0737805/README.md#compression)? I'm only asking because the latter seems like a better fit here, but happy to be corrected 🙂

Reply:
I'd prefer the former, since we're talking about varying the hash algo used by Partitioner, but not changing its fundamental strategy. Also, I prefer a minimal change in API – the high level libraries can easily make this a config key without any extra APIs.

Reply:
Great, thanks for breaking it down for me. I've removed the API change and moved it inside the `Partitioner`